src/Host/Broker/Impl/RemoteUri/RemoteUriHelper.cs

```diff
@@ -5,37 +5,88 @@
+                if (ct.IsCancellationRequested) {
+                    Task.WhenAll(
+                        receiver.CloseAsync(WebSocketCloseStatus.NormalClosure, "Normal closure", CancellationToken.None),
+                        sender.CloseAsync(WebSocketCloseStatus.NormalClosure, "Normal closure", CancellationToken.None))
+                        .DoNotWait();
+                    return;
+                }
+
+                WebSocketReceiveResult result = await receiver.ReceiveAsync(receiveBuffer, ct);
+
+                byte[] data = null;
+                using (MemoryStream ms = new MemoryStream()) {
+                    await ms.WriteAsync(receiveBuffer.Array, receiveBuffer.Offset, result.Count);
+                    await ms.FlushAsync();
+                    data = ms.ToArray();
+                }
+
+                ArraySegment<byte> sendBuffer = new ArraySegment<byte>(data);
+                await sender.SendAsync(sendBuffer, result.MessageType, result.EndOfMessage, ct);
+
+                if (result.MessageType == WebSocketMessageType.Close) {
+                    await receiver.CloseAsync(WebSocketCloseStatus.NormalClosure, "Normal closure", ct);
+                }
+            }
+        }
+
+        private static Task DoWebSocketWorkAsync(WebSocket localWebSocket, WebSocket remoteWebSocket, CancellationToken ct) {
+            return Task.WhenAll(
+                DoWebSocketReceiveSendAsync(localWebSocket, remoteWebSocket, ct),
+                DoWebSocketReceiveSendAsync(remoteWebSocket, localWebSocket, ct));
+        }
+
         public static async Task HandlerAsync(HttpContext context) {
             var url = context.Request.Headers[CustomHttpHeaders.RTVSRequestedURL];
 
             HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
             request.Method = context.Request.Method;
-            SetRequestHeaders(request, context.Request.Headers);
 
-            if (context.Request.ContentLength > 0) {
-                using (Stream reqStream = await request.GetRequestStreamAsync()) {
-                    await context.Request.Body.CopyToAsync(reqStream);
-                    await reqStream.FlushAsync();
+            ClientWebSocket clientWebsocket = null;
+            if (context.WebSockets.IsWebSocketRequest) {
+                UriBuilder ub = new UriBuilder(url) { Scheme = "ws" };
+                clientWebsocket = new ClientWebSocket();
+                await clientWebsocket.ConnectAsync(ub.Uri, CancellationToken.None);
+                var serverWebSocket = await context.WebSockets.AcceptWebSocketAsync(clientWebsocket.SubProtocol);
+                await DoWebSocketWorkAsync(serverWebSocket, clientWebsocket, CancellationToken.None);
+            } else {
+                SetRequestHeaders(request, context.Request.Headers);
+
+                if (context.Request.ContentLength > 0) {
+                    using (Stream reqStream = await request.GetRequestStreamAsync()) {
```

Comment:
It should be very uncommon to work with request stream directly - normally you'd have a controller and let MVC take care of JSON serialization.